pkg/skaffold/kubernetes/port_forward.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2018 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os/exec"
+	"strconv"
+	"sync"
+	"syscall"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/color"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+// PortForwarder is responsible for selecting pods satisfying a certain condition and port-forwarding the exposed
+// container ports within those pods. It also tracks and manages the port-forward connections.
+type PortForwarder struct {
```

Comment:
nit: I would rename this to `ExclusivePortForwarder` (or similar) and rename the `Forwarder` interface to `PortForwarder`.